### PR TITLE
Utkarsh4321/tutorial binary search tree

### PR DESCRIPTION
**Pull request by utkarsh4321**
## Change Summary
Issue #62  
In this PR I discuss BST.
Talk about the basic operations in BST.
Like search, deletion and insertion.
Also, talk about the Time complexity of these operations.
And in the end, I also suggested some problems for BST.

## Checklist

**If you haven't fulfilled the below requirements, your PR won't be reviewed.**

### General 

- [x] Make sure you've read [CONTRIBUTING.md](https://github.com/wingkwong/leetcode-the-hard-way/blob/main/CONTRIBUTING.md) 
- [x] Make sure you've followed the template format
- [x] Make sure you've started the app locally and verified all the content and all links (if applicable) must be accessible correctly
- [x] Make sure you've applied LaTex for variables, formulas and time / space complexity instead of using backticks
- [x] Make sure you've formatted your code well (Code Block is rendered using `<Tabs/>`)
      Below is a C++ example. You can include multiple langs using `<TabItem/>` if necessary
      
````
<Tabs>

<TabItem value="cpp" label="C++">
<SolutionAuthor name="@YOUR_ALIAS"/>

```cpp
// Your code goes here
// Prefer K&R Coding style
```

</TabItem>

</Tabs>
````
- [x] Make sure you've explained your code well (write comments above each line - not inline)
- [x] Make sure you've written your explanation well and it is easy to understand for beginners
- [x] Make sure you've included Time & Space Complexity Analysis
- [ ] Make sure you've included a meaningful approach name for your solution. Don't leave it blank. e.g. `## Approach 1: Two Pointers`.

### For Tutorial Articles

- [x] Make sure you've explained your topic using 1 - 3 LC problems rather than using external problems. Solutions must be taken with explanations as well.
- [x] Make sure you've provided suggested problems at the end using the given format. See [here](https://raw.githubusercontent.com/wingkwong/leetcode-the-hard-way/main/tutorials/math/number-theory/binary-exponentiation.md) as an example. If the target solution is not available, leave `solutionLink` blank.
- [ ] Make sure you've given credits / references if you use external resources. For an image, give credit under it. Otherwise, add a new section called References at the end (after Suggested Problems).

### CODE REVIEW

**Review by wingkwong — CHANGES_REQUESTED** [review]
1. you are expected to tick all items in checklist. (Not only tick, but also verify them one by one before submitting)
2. most of your content is not fulfilling the checklist. (wrong code block format, not using LC problems as examples, inline comments etc)
3. adding some images will be better

**Comment by utkarsh4321** [comment]
@wingkwong hi I m not got to explain with 1 LC problem can you please tell? For BST in LC, there is some problems like search in BST or Insertion in BST so I will take those problems explain on the basis of it. Is there any article. That i can fallow.

What's wrong in code block?

**Comment by wingkwong** [comment]
1. include everything that a beginner should know. explained with LC problems.
2. For code block, please look at the checklist or other tutorials for reference.
image

**Comment by utkarsh4321** [comment]
@wingkwong please review my changes and let me know if there are any changes you want I will try to resolve it asap. Thankyou

File: tutorials/graph-theory/binary-search-tree.md [hunk 1]
```diff
@@ -1,12 +1,391 @@
 ---
-title: 'Binary Search Tree'
-description: 'Coming Soon!'
+title: "Binary Search Tree (BST)"
+description: "Author: @utkarsh"
```
Comment (wingkwong) [inline tutorials/graph-theory/binary-search-tree.md:3]: this's not description for this topic. it should be related to BST instead. check out other tutorials for reference.

File: tutorials/graph-theory/binary-search-tree.md [hunk 2]
```diff
@@ -1,12 +1,391 @@
...
+1. The left node of the tree is smaller (or equal) to the root node.
+2. The right node of the tree is greater (or equal) to the root node.
+3. All the nodes of left subtree(tree left side of root node) are smaller (or equal) to the root node.
+4. And All the nodes of right subtree (tree right side of the root node) are greater (or equal) to the root node.
+
+<!-- Diagram of BST -->
+
+![image](https://user-images.githubusercontent.com/30936607/195894102-11f3d201-e1ef-4fd7-83de-5e00e6d66eb7.png)
+
+![image](https://user-images.githubusercontent.com/30936607/195677540-23c39b0f-ce06-4c7d-bffa-468b6c1b3f79.png)
```
Comment (wingkwong) [inline tutorials/graph-theory/binary-search-tree.md:36]: did you make these yourself?

File: tutorials/graph-theory/binary-search-tree.md [hunk 3]
```diff
@@ -1,12 +1,391 @@
...
+
+## Overview
+
+Before jump to BST. We recap some basics to better understand BST:
+
+A tree is a data structure composed of nodes that has some following characteristics:
+
+1. Each tree has root node at the top (also known as Parent Node).
+2. The root node has zero or more child nodes.
+3. Each child node has zero or more child nodes. This create subtree in the tree. Every node has its own subtree made up of its children etc. This means that every node on its own can be tree.
```
Comment (wingkwong) [inline tutorials/graph-theory/binary-search-tree.md:23]: i believe these are copied from https://www.freecodecamp.org/news/binary-search-trees-bst-explained-with-examples/

File: tutorials/graph-theory/binary-search-tree.md [hunk 4]
```diff
@@ -1,12 +1,391 @@
...
+
+![image](https://user-images.githubusercontent.com/30936607/195894102-11f3d201-e1ef-4fd7-83de-5e00e6d66eb7.png)
+
+![image](https://user-images.githubusercontent.com/30936607/195677540-23c39b0f-ce06-4c7d-bffa-468b6c1b3f79.png)
+
+Like binary tree. We can also traverse BST in **pre**,**post** and **inorder** traversal.
+But the amazing fact is that **inorder** traversal
+of `BST` will give **sorted data**.
+
+### Inorder traversal in BST
```
Comment (wingkwong) [inline tutorials/graph-theory/binary-search-tree.md:42]: use LC problem

File: tutorials/graph-theory/binary-search-tree.md [hunk 5]
```diff
@@ -1,12 +1,391 @@
...
+<TabItem value="cpp" label="C++">
+<SolutionAuthor name="@utkarsh"/>
+
+```cpp
+// For BST search
+Node *bstSearch(int target,Node* root){
+    if(root == NULL)
+        return root;
+    // If target value matches with node value
+    if(root->data == target)c
```
Comment (wingkwong) [inline tutorials/graph-theory/binary-search-tree.md:121]: unexpected character `c`

File: tutorials/graph-theory/binary-search-tree.md [hunk 6]
```diff
@@ -1,12 +1,391 @@
...
+    // Return node
```
Comment (wingkwong) [inline tutorials/graph-theory/binary-search-tree.md:122]: this comment is useless

**Review by wingkwong — CHANGES_REQUESTED** [review]
1. Don't follow GFG style. It's really awful. 
2. See https://gist.github.com/jesseschalken/0f47a2b5a738ced9c845#why-kr for the coding style
3. The code you've provided is not tailored to LC problems at all. I'm expecting to see a LC solution that is using the snippets that you wrote. Not just posting a random c++ program here.

File: tutorials/graph-theory/binary-search-tree.md (hunk 2 above)
Comment (utkarsh4321) [reply tutorials/graph-theory/binary-search-tree.md:36]: yes

File: tutorials/graph-theory/binary-search-tree.md (hunk 3 above)
Comment (utkarsh4321) [reply tutorials/graph-theory/binary-search-tree.md:23]: yes, you want to remove it?

File: tutorials/graph-theory/binary-search-tree.md (hunk 5 above)
Comment (utkarsh4321) [reply tutorials/graph-theory/binary-search-tree.md:121]: ok i will remove it

File: tutorials/graph-theory/binary-search-tree.md (hunk 6 above)
Comment (utkarsh4321) [reply tutorials/graph-theory/binary-search-tree.md:122]: ok

File: tutorials/graph-theory/binary-search-tree.md (hunk 3 above)
Comment (utkarsh4321) [reply tutorials/graph-theory/binary-search-tree.md:23]: ok

File: tutorials/graph-theory/binary-search-tree.md (hunk 4 above)
Comment (utkarsh4321) [reply tutorials/graph-theory/binary-search-tree.md:42]: Here for Inorder traversal, you want LC Problem to explain the Inorder traversal of BST. So I need to search for BST in order in LC and that problem explanation you want

File: tutorials/graph-theory/binary-search-tree.md (hunk 1 above)
Comment (utkarsh4321) [reply tutorials/graph-theory/binary-search-tree.md]: @wingkwong  i added review it once if you want something else please tell me i will update it.

File: tutorials/graph-theory/binary-search-tree.md (hunk 2 above)
Comment (utkarsh4321) [reply tutorials/graph-theory/binary-search-tree.md:36]: yes, I made this with the help of canva @wingkwong.

File: tutorials/graph-theory/binary-search-tree.md (hunk 3 above)
Comment (utkarsh4321) [reply tutorials/graph-theory/binary-search-tree.md:23]: Added reference please check that

File: tutorials/graph-theory/binary-search-tree.md (hunk 5 above)
Comment (utkarsh4321) [reply tutorials/graph-theory/binary-search-tree.md]: remote that please check it

File: tutorials/graph-theory/binary-search-tree.md (hunk 6 above)
Comment (utkarsh4321) [reply tutorials/graph-theory/binary-search-tree.md]: remove these comments.

**Comment by utkarsh4321** [comment]
@wingkwong please check that I did some minor changes. If you are ok with that then I move further. If you want something else tell me I will try to fix it asap.

File: tutorials/graph-theory/binary-search-tree.md [hunk 7]
```diff
@@ -1,12 +1,304 @@
...
+1. The left node of the tree is smaller (or equal) to the root node.
+2. The right node of the tree is greater (or equal) to the root node.
+3. All the nodes of left subtree(tree left side of root node) are smaller (or equal) to the root node.
+4. And All the nodes of right subtree (tree right side of the root node) are greater (or equal) to the root node.
+
+<!-- Diagram of BST -->
+
+![image](https://user-images.githubusercontent.com/30936607/195894102-11f3d201-e1ef-4fd7-83de-5e00e6d66eb7.png)
+
+![image](https://user-images.githubusercontent.com/30936607/195677540-23c39b0f-ce06-4c7d-bffa-468b6c1b3f79.png)
```
Comment (wingkwong) [inline tutorials/graph-theory/binary-search-tree.md:36]: suggest to draw better circles

File: tutorials/graph-theory/binary-search-tree.md [hunk 8]
```diff
@@ -1,12 +1,304 @@
...
+A valid **binary search tree** (BST) is defined as follows:
+
+1. The left node of the tree is smaller (or equal) to the root node.
+2. The right node of the tree is greater (or equal) to the root node.
+3. All the nodes of left subtree(tree left side of root node) are smaller (or equal) to the root node.
+4. And All the nodes of right subtree (tree right side of the root node) are greater (or equal) to the root node.
+
+<!-- Diagram of BST -->
+
+![image](https://user-images.githubusercontent.com/30936607/195894102-11f3d201-e1ef-4fd7-83de-5e00e6d66eb7.png)
```
Comment (wingkwong) [inline tutorials/graph-theory/binary-search-tree.md:34]: Bst -> BST

File: tutorials/graph-theory/binary-search-tree.md [hunk 9]
```diff
@@ -1,12 +1,304 @@
...
+But the amazing fact is that **inorder** traversal
+of `BST` will give **sorted data**.
+
+### Inorder traversal in BST
+
+<Tabs>
+<TabItem value="cpp" label="C++">
+<SolutionAuthor name="@utkarsh"/>
+
+```cpp
```
Comment (wingkwong) [inline tutorials/graph-theory/binary-search-tree.md:48]: Format your code with K&R style.

See https://gist.github.com/jesseschalken/0f47a2b5a738ced9c845#why-kr for example

File: tutorials/graph-theory/binary-search-tree.md [hunk 10]
```diff
@@ -1,12 +1,304 @@
...
+4. If value not matched with current node then we either procced to right or left child according to the condition.
+5. If target node value is greater than the current node then we procced towards the right child or right subtree.
+6. If target node value is lower than the current node then we proceed towards the left child or left subtree.
+
+`target value: value of node you want to search in tree`
+<Tabs>
+<TabItem value="cpp" label="C++">
+<SolutionAuthor name="@utkarsh"/>
+
+```cpp
```
Comment (wingkwong) [inline tutorials/graph-theory/binary-search-tree.md:115]: Format your code with K&R style.

See https://gist.github.com/jesseschalken/0f47a2b5a738ced9c845#why-kr for example

File: tutorials/graph-theory/binary-search-tree.md [hunk 11]
```diff
@@ -1,12 +1,304 @@
...
+1. Insertion begin with searh operation.
+2. Search left or right subtree according to the relation between node value and the value of our target node as we did in search.
+3. Repeat step 1 until we reach to correct place to insert our new node.
+4. Add a new node as a right or left child according to the node value and our target node value.
+
+<Tabs>
+<TabItem value="cpp" label="C++">
+<SolutionAuthor name="@utkarsh"/>
+
+```cpp
```
Comment (wingkwong) [inline tutorials/graph-theory/binary-search-tree.md:148]: Format your code with K&R style.

See https://gist.github.com/jesseschalken/0f47a2b5a738ced9c845#why-kr for example

File: tutorials/graph-theory/binary-search-tree.md [hunk 12]
```diff
@@ -1,12 +1,304 @@
...
+- Find the inorder successor of that node. (Inorder successor:- the leftmost node in the right subtree).
+- Replace that node with inorder successor.
+- Delete inorder successor from it's original position.
+
+<Tabs>
+<TabItem value="cpp" label="C++">
+<SolutionAuthor name="@utkarsh"/>
+
+```cpp
+Node *inorderSuccessor(Node* root){
```
Comment (wingkwong) [inline tutorials/graph-theory/binary-search-tree.md:201]: Format your code with K&R style.

See https://gist.github.com/jesseschalken/0f47a2b5a738ced9c845#why-kr for example

File: tutorials/graph-theory/binary-search-tree.md [hunk 13]
```diff
@@ -1,12 +1,304 @@
...
+"leetCodeLink": "https://leetcode.com/problems/range-sum-of-bst/",
+"solutionLink": ""
+},
+{
+"problemName": "098 - Validate Binary Search Tree",
+"difficulty": "Medium",
+"leetCodeLink": "https://leetcode.com/problems/validate-binary-search-tree/",
+"solutionLink": ""
+}
+]
```
Comment (wingkwong) [inline tutorials/graph-theory/binary-search-tree.md:296]: format it well

File: tutorials/graph-theory/binary-search-tree.md [hunk 14]
```diff
@@ -1,12 +1,304 @@
...
+3. All the nodes of left subtree(tree left side of root node) are smaller (or equal) to the root node.
+4. And All the nodes of right subtree (tree right side of the root node) are greater (or equal) to the root node.
+
+<!-- Diagram of BST -->
+
+![image](https://user-images.githubusercontent.com/30936607/195894102-11f3d201-e1ef-4fd7-83de-5e00e6d66eb7.png)
+
+![image](https://user-images.githubusercontent.com/30936607/195677540-23c39b0f-ce06-4c7d-bffa-468b6c1b3f79.png)
+
+Like binary tree. We can also traverse BST in **pre**,**post** and **inorder** traversal.
```
Comment (wingkwong) [inline tutorials/graph-theory/binary-search-tree.md:38]: you may mention the difference between binary tree & binary search tree.

File: tutorials/graph-theory/binary-search-tree.md [hunk 15]
```diff
@@ -1,12 +1,304 @@
...
+But the amazing fact is that **inorder** traversal
+of `BST` will give **sorted data**.
```
Comment (wingkwong) [inline tutorials/graph-theory/binary-search-tree.md:40]: unexpected line break

**Review by wingkwong — CHANGES_REQUESTED** [review]

**Comment by wingkwong** [comment]
Closing due to inactivity.

**Pull request closed without merging**